// hazelcast-client/lib/serialization/PredicateFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PREDICATE_FACTORY_ID = -32;
var AbstractPredicate = (function () {
    function AbstractPredicate() {
    }
    AbstractPredicate.prototype.getFactoryId = function () {
        return exports.PREDICATE_FACTORY_ID;
    };
    return AbstractPredicate;
}());
exports.AbstractPredicate = AbstractPredicate;
var PredicateFactory = (function () {
    function PredicateFactory(allPredicates) {
        this.idToConstructorMap = {};
        for (var pred in allPredicates) {
            this.idToConstructorMap[allPredicates[pred].prototype.getClassId()] = allPredicates[pred];
        }
    }
    PredicateFactory.prototype.create = function (type) {
        if (this.idToConstructorMap[type]) {
            return (new this.idToConstructorMap[type]());
        }
        else {
            throw new RangeError("There is no default predicate with id " + type + ".");
        }
    };
    return PredicateFactory;
}());
exports.PredicateFactory = PredicateFactory;
