// hazelcast-client/lib/Address.js
"use strict";
var ClusterDataFactory_1 = require("./ClusterDataFactory");
var net = require("net");
var Address = (function () {
    function Address(host, port) {
        this.host = host;
        this.port = port;
        if (net.isIPv6(host)) {
            this.type = 6;
        }
        else {
            this.type = 4;
        }
    }
    Address.prototype.readData = function (input) {
        this.port = input.readInt();
        this.type = input.readByte();
        this.host = input.readUTF();
    };
    Address.prototype.writeData = function (output) {
        output.writeInt(this.port);
        output.writeByte(this.type);
        output.writeUTF(this.host);
    };
    Address.prototype.getFactoryId = function () {
        return ClusterDataFactory_1.CLUSTER_DATA_FACTORY_ID;
    };
    Address.prototype.getClassId = function () {
        return ClusterDataFactory_1.ADDRESS_CLASS_ID;
    };
    Address.prototype.toString = function () {
        return this.host + ':' + this.port;
    };
    return Address;
}());
module.exports = Address;
