// hazelcast-client/lib/PartitionService.js
"use strict";
var Promise = require("bluebird");
var GetPartitionsCodec = require("./codec/GetPartitionsCodec");
var PARTITION_REFRESH_INTERVAL = 10000;
var PartitionService = (function () {
    function PartitionService(client) {
        this.partitionMap = {};
        this.client = client;
        this.isShutdown = false;
    }
    PartitionService.prototype.initialize = function () {
        return this.refresh();
    };
    PartitionService.prototype.shutdown = function () {
        this.isShutdown = true;
        clearTimeout(this.partitionRefreshTask);
    };
    PartitionService.prototype.refresh = function () {
        var _this = this;
        if (this.isShutdown) {
            return Promise.resolve();
        }
        var ownerConnection = this.client.getClusterService().getOwnerConnection();
        var clientMessage = GetPartitionsCodec.encodeRequest();
        return this.client.getInvocationService()
            .invokeOnConnection(ownerConnection, clientMessage)
            .then(function (clientMessage) {
            var receivedPartitionMap = GetPartitionsCodec.decodeResponse(clientMessage);
            for (var partitionId in receivedPartitionMap) {
                _this.partitionMap[partitionId] = receivedPartitionMap[partitionId];
            }
            _this.partitionCount = Object.keys(_this.partitionMap).length;
        }).finally(function () {
            _this.partitionRefreshTask = setTimeout(_this.refresh.bind(_this), PARTITION_REFRESH_INTERVAL);
        });
    };
    PartitionService.prototype.getAddressForPartition = function (partitionId) {
        return this.partitionMap[partitionId];
    };
    PartitionService.prototype.getPartitionId = function (key) {
        var partitionHash;
        if (typeof key === 'object' && 'getPartitionHash' in key) {
            partitionHash = key.getPartitionHash();
        }
        else {
            partitionHash = this.client.getSerializationService().toData(key).getPartitionHash();
        }
        return Math.abs(partitionHash) % this.partitionCount;
    };
    PartitionService.prototype.getPartitionCount = function () {
        return this.partitionCount;
    };
    return PartitionService;
}());
module.exports = PartitionService;
