// hazelcast-client/lib/proxy/topic/RawTopicMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RELIABLE_TOPIC_MESSAGE_FACTORY_ID = -18;
exports.RELIABLE_TOPIC_CLASS_ID = 2;
var RawTopicMessage = (function () {
    function RawTopicMessage() {
    }
    RawTopicMessage.prototype.readData = function (input) {
        this.publishTime = input.readLong();
        this.publisherAddress = input.readObject();
        this.payload = input.readData();
    };
    RawTopicMessage.prototype.writeData = function (output) {
        output.writeLong(this.publishTime);
        output.writeObject(this.publisherAddress);
        output.writeData(this.payload);
    };
    RawTopicMessage.prototype.getFactoryId = function () {
        return exports.RELIABLE_TOPIC_MESSAGE_FACTORY_ID;
    };
    RawTopicMessage.prototype.getClassId = function () {
        return exports.RELIABLE_TOPIC_CLASS_ID;
    };
    return RawTopicMessage;
}());
exports.RawTopicMessage = RawTopicMessage;
var ReliableTopicMessageFactory = (function () {
    function ReliableTopicMessageFactory() {
    }
    ReliableTopicMessageFactory.prototype.create = function (type) {
        if (type === exports.RELIABLE_TOPIC_CLASS_ID) {
            return new RawTopicMessage();
        }
        return null;
    };
    return ReliableTopicMessageFactory;
}());
exports.ReliableTopicMessageFactory = ReliableTopicMessageFactory;
