// hazelcast-client/lib/invocation/RegistrationKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RegistrationKey = (function () {
    function RegistrationKey(regId, codec, registerRequest, registerHandlerFunc) {
        this.userRegistrationId = regId;
        this.registerHandlerFunc = registerHandlerFunc;
        this.registerRequest = registerRequest;
        this.codec = codec;
    }
    RegistrationKey.prototype.getRegisterRequest = function () {
        return this.registerRequest;
    };
    RegistrationKey.prototype.setRegisterRequest = function (registerRequest) {
        this.registerRequest = registerRequest;
    };
    RegistrationKey.prototype.getCodec = function () {
        return this.codec;
    };
    RegistrationKey.prototype.setCodec = function (value) {
        this.codec = value;
    };
    RegistrationKey.prototype.getHandler = function () {
        return this.registerHandlerFunc;
    };
    RegistrationKey.prototype.setHandler = function (handler) {
        this.registerHandlerFunc = handler;
    };
    RegistrationKey.prototype.getUserRegistrationKey = function () {
        return this.userRegistrationId;
    };
    return RegistrationKey;
}());
exports.RegistrationKey = RegistrationKey;
